egs/wsj/s5/steps/data/reverberate_data_dir.py

```diff
@@ -0,0 +1,506 @@
+    f = open(output_file, "w")
+    for i in range(1, num_replicas+1):
+        for line in list:
+            if len(line) > 0 and line[0] != ';':
+                split1 = line.split()
+                for j in field:
+                    split1[j] = GetNewId(split1[j], prefix, i)
+                print(" ".join(split1), file=f)
+            else:
+                print(line, file=f)
+    f.close()
+
+
+# This function creates multiple copies of the necessary files, e.g. utt2spk, wav.scp ...
+def CreateReverberatedCopy(input_dir,
+                           output_dir,
+                           room_dict,  # the room dictionary, please refer to MakeRoomDict() for the format
+                           pointsource_noise_list, # the point source noise list
+                           iso_noise_list, # the isotropic noise list
+                           foreground_snr_string, # the SNR for adding the foreground noises
+                           background_snr_string, # the SNR for adding the background noises
+                           num_replicas, # Number of replicate to generated for the data
+                           prefix, # prefix for the id of the corrupted utterances
+                           speech_rvb_probability, # Probability of reverberating a speech signal
+                           isotropic_noise_addition_probability, # Probability of adding isotropic noises
+                           pointsource_noise_addition_probability, # Probability of adding point-source noises
+                           max_noises_per_minute  # maximum number of point-source noises that can be added to a recording according to its duration
+                           ):
+    
+    if not os.path.isfile(input_dir + "/reco2dur"):
+        print("Getting the duration of the recordings...");
+        data_lib.RunKaldiCommand("wav-to-duration --read-entire-file=true scp:{0}/wav.scp ark,t:{0}/reco2dur".format(input_dir))
+    durations = ParseFileToDict(input_dir + "/reco2dur", value_processor = lambda x: float(x[0]))
+    wav_scp = ParseFileToDict(input_dir + "/wav.scp", value_processor = lambda x: " ".join(x))
+    foreground_snr_array = map(lambda x: float(x), foreground_snr_string.split(':'))
+    background_snr_array = map(lambda x: float(x), background_snr_string.split(':'))
+
+    GenerateReverberatedWavScp(wav_scp, durations, output_dir, room_dict, pointsource_noise_list, iso_noise_list, 
+               foreground_snr_array, background_snr_array, num_replicas, prefix, 
+               speech_rvb_probability, isotropic_noise_addition_probability, 
+               pointsource_noise_addition_probability, max_noises_per_minute)
+
+    AddPrefixToFields(input_dir + "/utt2spk", output_dir + "/utt2spk", num_replicas, prefix, field = [0,1])
+    data_lib.RunKaldiCommand("utils/utt2spk_to_spk2utt.pl <{output_dir}/utt2spk >{output_dir}/spk2utt"
+                    .format(output_dir = output_dir))
+
+    if os.path.isfile(input_dir + "/text"):
```

Comment:
```bash
 grep -R steps -e utt2uniq
steps/cleanup/combine_short_segments.py:    utt2uniq = None
steps/cleanup/combine_short_segments.py:    if os.path.exists(data_dir_file('utt2uniq')):
steps/cleanup/combine_short_segments.py:        utt2uniq = ParseFileToDict(data_dir_file('utt2uniq'))
steps/cleanup/combine_short_segments.py:    return utt2spk, spk2utt, text, feat, utt2dur, utt2uniq
steps/cleanup/combine_short_segments.py:def WriteCombinedDirFiles(output_dir, utt2spk, spk2utt, text, feat, utt2dur, utt2uniq):
steps/cleanup/combine_short_segments.py:        # updating utt2uniq
steps/cleanup/combine_short_segments.py:        if utt2uniq is not None:
steps/cleanup/combine_short_segments.py:                combined_uniqs.append(utt2uniq.pop(utt))
steps/cleanup/combine_short_segments.py:            # utt2uniq file is used to map perturbed data to original unperturbed
steps/cleanup/combine_short_segments.py:            # is no good way to find the utt2uniq mappinng so that we can avoid
steps/cleanup/combine_short_segments.py:            utt2uniq[new_utt_name] = combined_uniqs[0]
steps/cleanup/combine_short_segments.py:    if utt2uniq is not None:
steps/cleanup/combine_short_segments.py:        WriteDictToFile(utt2uniq, out_dir_file('utt2uniq'))
steps/cleanup/combine_short_segments.py:    utt2spk, spk2utt, text, feat, utt2dur, utt2uniq = ParseDataDirInfo(input_dir)
steps/cleanup/combine_short_segments.py:    WriteCombinedDirFiles(output_dir, utt2spk, spk2utt, text, feat, utt2dur, utt2uniq)
steps/nnet2/get_egs.sh:if [ -f $data/utt2uniq ]; then
steps/nnet2/get_egs.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet2/get_egs.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet2/get_egs.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet2/get_perturbed_feats.sh:# In the combined feature directory, create a file utt2uniq which maps
steps/nnet2/get_perturbed_feats.sh:  > $data/utt2uniq
steps/nnet2/get_egs2.sh:if [ -f $data/utt2uniq ]; then
steps/nnet2/get_egs2.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet2/get_egs2.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet2/get_egs2.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet3/chain/get_egs.sh:if [ -f $data/utt2uniq ]; then  # this matters if you use data augmentation.
steps/nnet3/chain/get_egs.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet3/chain/get_egs.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet3/chain/get_egs.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet3/get_egs_targets.sh:if [ -f $data/utt2uniq ]; then  # this matters if you use data augmentation.
steps/nnet3/get_egs_targets.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet3/get_egs_targets.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet3/get_egs_targets.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet3/get_egs.sh:if [ -f $data/utt2uniq ]; then  # this matters if you use data augmentation.
steps/nnet3/get_egs.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet3/get_egs.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet3/get_egs.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/nnet3/get_egs_discriminative.sh:if [ -f $data/utt2uniq ]; then  # this matters if you use data augmentation.
steps/nnet3/get_egs_discriminative.sh:  echo "File $data/utt2uniq exists, so augmenting valid_uttlist to"
steps/nnet3/get_egs_discriminative.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/nnet3/get_egs_discriminative.sh:  cat $dir/valid_uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/online/nnet2/get_egs.sh:if [ -f $data/utt2uniq ]; then
steps/online/nnet2/get_egs.sh:  echo "File $data/utt2uniq exists, so augmenting valid/uttlist to"
steps/online/nnet2/get_egs.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/online/nnet2/get_egs.sh:  cat $dir/valid/uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
steps/online/nnet2/get_egs2.sh:if [ -f $data/utt2uniq ]; then
steps/online/nnet2/get_egs2.sh:  echo "File $data/utt2uniq exists, so augmenting valid/uttlist to"
steps/online/nnet2/get_egs2.sh:  utils/utt2spk_to_spk2utt.pl $data/utt2uniq > $dir/uniq2utt
steps/online/nnet2/get_egs2.sh:  cat $dir/valid/uttlist.tmp | utils/apply_map.pl $data/utt2uniq | \
```